src/Entity/SmsMessage.php

```diff
@@ -134,25 +135,73 @@ public function removeOption($name) {
    * {@inheritdoc}
    */
   public function getResult() {
-    return $this->get('result')->entity;
+    $results = $this->entityTypeManager()
+      ->getStorage('sms_result')
+      ->loadByProperties(['parent' => $this->id()]);
+    return $results ? reset($results) : NULL;
   }
 
   /**
    * {@inheritdoc}
    */
   public function setResult(StdMessageResultInterface $result = NULL) {
     if ($result === NULL) {
-      $this->set('result', NULL);
+      // Find any pre-existing report and remove the parent-child relationship.
+      if ($previous_result = $this->getResult()) {
```

Comment:
The main dilemma here is that the `setResult()` operation doesn't commit to storage until `->save()` is called. Therefore, deleting at this point may be premature because `->save()` may never be called, yet we have already committed changes.

This in fact, is the key challenge with the back-referencing approach - the fact that you have to save changes to the child when the parent hasn't been saved.